tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -12,90 +12,457 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 
 #include "tensorflow/compiler/xla/service/gpu/target_util.h"
 
+#include "absl/strings/str_cat.h"
 #include "llvm/IR/MDBuilder.h"
+#include "tensorflow/compiler/xla/primitive_util.h"
+#include "tensorflow/compiler/xla/service/llvm_ir/ir_builder_mixin.h"
 #include "tensorflow/compiler/xla/service/llvm_ir/llvm_util.h"
+#include "tensorflow/compiler/xla/xla_data.pb.h"
 #include "tensorflow/core/platform/logging.h"
 
 namespace xla {
+
+using absl::StrAppend;
+
 namespace gpu {
 namespace {
+
 // Utility functions to obtain NVPTX/AMDGPU specific information.
 
-// Wrapper structure for carrying llvm intrinsic ids for NVPTX/AMDGPU platforms.
-struct TargetIntrinsics {
-  llvm::Intrinsic::ID nvptx_intrinsic;
-  llvm::Intrinsic::ID amdgpu_intrinsic;
+// Wrapper structure for  intrinsic information for NVPTX/AMDGPU.
+struct TargetIntrinsicInfo {
+  TargetIntrinsicInfo(llvm::Intrinsic::ID x) : intrinsic(x) {}
+  llvm::Intrinsic::ID intrinsic;
+};
+
+// Wrapper structure for  device functionc information for NVPTX/AMDGPU.
+struct TargetFunctionInfo {
+  // No signature mismatch between call site and device function prototype.
+  TargetFunctionInfo(const string callee_name_, bool check_signature_)
+      : callee_name(callee_name_),
+        check_signature(check_signature_),
+        input_types({PRIMITIVE_TYPE_INVALID}),
+        result_type(PRIMITIVE_TYPE_INVALID) {}
+  // Potential signature mismatch - generate conversions if needed.
+  TargetFunctionInfo(const string callee_name_, bool check_signature_,
+                     absl::Span<const PrimitiveType> input_types_,
+                     const PrimitiveType output_type_)
+      : callee_name(callee_name_),
+        check_signature(check_signature_),
+        input_types(input_types_),
+        result_type(output_type_) {}
+  // Potential signature mistamatch - use bitcast if specified by the target.
+  TargetFunctionInfo(const string callee_name_, bool check_signature_,
+                     absl::Span<const PrimitiveType> input_types_,
+                     const PrimitiveType output_type_, bool use_bitcast_)
+      : callee_name(callee_name_),
+        check_signature(check_signature_),
+        input_types(input_types_),
+        result_type(output_type_),
+        use_bitcast(use_bitcast_) {}
+  // Device function name.
+  const string callee_name;
+  // Check signature of the device function and if needed generate casts.
+  bool check_signature;
+  // Inpute types accespted by the device function.
+  absl::Span<const PrimitiveType> input_types;
+  // Result type of the device function.
+  PrimitiveType result_type;
+  // Use bitcast to generate casts if the desired signature at the call site
+  // does not match the signature of the device function.
+  absl::optional<bool> use_bitcast;
+};
+
+// Wrapper structure to carry either information about the intrinsic
+// or device function for NVPTX/AMDGPU.
+// Note: Only intrinsic or device function - not both - is expected.
+struct TargetInfo {
```

Comment:
"TargetInfo" is not a good name for this struct.  The name suggests it contains info about a target, but actually this contains info about a call in a target, be it a function call or an intrinsic call.  But maybe we can just replace this by a variant type and remove this struct altogether.

Reply:
Am calling it TargetFunctionInfo that populates the variants for PTX and AMDGPU